File: tree.go
package main

import "fmt"

type node interface {
	write()
	children() []node
}
type baseNode struct {
	sub []node
}

func (self baseNode) children() []node {
	return self.sub
}

func subwrite(n node) {
	n.write()
	for _, sub := range n.children() {
		subwrite(sub)
	}
}

type identWriter struct {
	baseNode
	name string
}

func (self identWriter) write() {
	fmt.Print(self.name)
}

func main() {
	writer := identWriter{
		name: "Hello, ",
		baseNode: identWriter{sub: []node{
			writer2{
				number:   2,
				baseNode: baseNode{sub: []node{}},
			},
		}},
	}
	subwrite(writer)
}

File: tree.rs
trait Writer {
    fn write(&self);
}

struct Node<T: Writer> {
    writer: T,
    sub: Vec<Node>,
}
impl<T: Writer> Node<T> {
    fn write(&self) {
        for child in &self.sub {
            self.write();
        }
        self.writer.write();
    }
}

struct IdentWriter {
    name: &'static str
}
impl Writer for IdentWriter {
    fn write(&self) {
        println!("{}", self.name);
    }
}

fn main() {
    let piss = Node{
        writer: IdentWriter{ name: "hello, " },
        sub: vec!(Node{
            writer: IdentWriter{ name: "world" },
            sub: vec!(),
        })
    };

    piss.write();
}
